bnb_yaz/new_bnb_downloader.py

```diff
@@ -0,0 +1,450 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+#
+# A tool for the automation of MARC downloads from the BNB.
+# This version reads BNB IDs from a PDF file and downloads records from the BNB server.
+# Author: Steven Lolong (steven.lolong@uni-tuebingen.de)
+# First created: November 2025
+# Usage: new_bnb_downloader.py <email> <working_directory>
+# Requires 'pdftotext' utility to parse PDF files.
+# The working directory should have the following structure:
+# The 'input' directory should contain the PDF files with BNB IDs.
+# The 'loaded' directory will store processed PDF files and successfully uploaded MARC files.
+# The 'marc' directory will store downloaded MARC files.
+# The 'logs' directory will store log files.
+
+
+
+import pexpect
+import re
+import os
+import sys
+import sqlite3
+import datetime
+# import bsz_util
+
+# Global variables
+db_connection : sqlite3.Connection
+bnb_yaz_client : pexpect.spawn
+pdf_file_name : str
+marc_output_file : str
+tmp_parse_file : str
+working_directory : str
+email_recipient : str
+total_new_extracted_ids : int = 0
+total_found_ids : int = 0
+total_not_found_ids : int = 0
+total_existing_ids : int = 0
+pdf_files : list = []
+DATABASE_NAME = "/usr/local/var/lib/tuelib/bnb_downloads.db"
+LOG_FILE = "log_bnb_downloader_" + datetime.datetime.now().strftime("%Y%m%d_%H%M%S") + ".log"
+
+
+
+# Logging function to log messages (INFO, ERROR, etc.) to a log file
+# level: 1 = INFO, 2 = WARNING, 3 = ERROR
+def Logging(message, level=1):
+    LEVEL_INFO = {1: "INFO", 2: "WARNING", 3: "ERROR"}
+    with open(working_directory + "logs/" + LOG_FILE, "a") as log_file:
+        log_file.write(LEVEL_INFO.get(level, "INFO") + " -- " + datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S") + " - " + message + "\n")
+
+
+# Initialize the SQLite database and create necessary tables if they do not exist
+def InitializeDatabase():
+    UNSUCCESSFUL_DOWNLOAD_TABLE_SCHEMA = '''CREATE TABLE IF NOT EXISTS unsuccessful_download
```

Comment:
Instead of using two tables with a very similar structure here, i would rather suggest merging them into one:

```
TABLE_SCHEMA = '''
CREATE TABLE IF NOT EXISTS records (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        external_id TEXT NOT NULL UNIQUE,
        source_file_name TEXT NOT NULL,
        date_added TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
        date_delivered TIMESTAMP DEFAULT NULL
)'''
```

You can query all IDs that still need to be downloaded by SELECTing 'date_delivered IS NULL', and then just set date_delivered when the dataset has been delivered successfully + to NOW().